tests/csep1_tests.py

```diff
@@ -0,0 +1,305 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+"""
+Created on Thu Jan 23 20:06:58 2020
+
+
+@author: khawaja
+"""
+import numpy
+import scipy.stats
+#import matplotlib.pyplot as plt
+
+def poisson_log_likelihood(observation, forecast):      #This calls Poisson PMF function
+    """Wrapper around scipy to compute the Poisson log-likelihood
+    
+    Args:
+        observation: Observed (Grided) seismicity
+        forecast: Forecast of a Model (Grided)
+    
+    Returns:
+        Log-Liklihood values of between binned-observations and binned-forecasts
+    """
+    return numpy.log(scipy.stats.poisson.pmf(observation, forecast))
+
+
+def forecast_realization_inverse_cdf(random_matrix, forecast):
+    """Wrapper around scipy inverse poisson cdf to compute new forecast using 
+        actual forecast and random numbers between 0 and 1
+    
+    Args:
+        random_matrix: Matrix of dimenions equal to forecast, containing random
+                       numbers between 0 and 1.
+        forecast: Actual Forecast of a Model (Grided)
+    
+    Returns:
+        A new forecast which is the realization of Actual forecast
+    """
+        
+    return scipy.stats.poisson.ppf(random_matrix, forecast) 
+
+
+def l_test(observation, forecast, num_simulations, model_name):
```

Comment:
this doesn't look to be the optimized likelihood function. this is ok for now, but we will want to change that

Reply:
Actually, 
I looked into the implementation of the optimized version of L test, implemented in MATLAB. I could reproduce that easily, but to be honest that was not appealing to my understanding, therefore I did not implement that. Let me walk you through that piece of code here, then we can decide what to do about it. 

According to my understanding, that implementing is using two loops. Furthermore, the method by which forecasat realization is computed does not sound interesting at all. See the steps below and then see the corresponding piece of code below: 

Step 1- for loop explicitly used
Step 2- Using "find" function of Matlab to look for certain criterion
Step 3- Then computing the realization of forecast by just adding +1 value to the actual forecast (means, no inverse Poisson CDF)
Step 4-  Compute Likelihood of "actual_forecast_value+1" AND "actual_forecast_value". 
Step 5- After loops end, sum all likelihoods. 
----------------Code Lines-------------
vRandom = pythonCSEPRandom(nNumberEvents, 1); 
vRatesCDF = [cumsum(vForecast(:, Forecast_Rate)./fTotalForecastRate)];

for nCnt = 1:nNumberEvents     (step 1)
  nIndex = min(find((vRandom(nCnt) < vRatesCDF)));       (step 2)
  vForecast(nIndex, Forecast_Observations) = vForecast(nIndex, Forecast_Observations) + 1;  (step 3)
  vForecast(nIndex, Forecast_PrecomputedZeroLikelihood) = calc_logpoisspdf(vForecast(nIndex, Forecast_Observations), vForecast(nIndex, Forecast_Rate)); (   step 4)
end;
fLikelihood = sum(vForecast(:, Forecast_PrecomputedZeroLikelihood)); (step 5)
--------------Code Ends-----------------------
This does not make sense all to me. If I am wrong, please guide me. 

However, I have found another implementation of Likelihood:
    Rhoades et al. 2011. (Equation 4): 
             A revised equation of LogLikelihood formula is provided. Which I can be a bit efficient, in a way that it skips the use of a Log and a Factorial. Rest of all steps to be remain same. 

Please enlighten me, if I am not getting it right.
Regards.

Reply:
Actually, 
I looked into the implementation of the optimized version of L test, implemented in MATLAB. I could reproduce that easily, but to be honest that was not appealing to my understanding, therefore I did not implement that. Let me walk you through that piece of code here, then we can decide what to do about it. 

According to my understanding, that implementation is using two loops (so not much faster). Furthermore, the method by which forecast realization is computed does not appeal to my understanding. See the steps below and then match the corresponding piece of code below: 

Step 1- for loop explicitly used
Step 2- Using "find" function of Matlab to look for certain criterion
Step 3- Then computing the realization of forecast by just adding +1 value to the actual forecast (means, no inverse Poisson CDF)
Step 4-  Compute Likelihood of "actual_forecast_rate+1" AND "actual_forecast_rate". 
Step 5- After loops end, sum all likelihoods. 
----------------Code Lines-------------
vRandom = pythonCSEPRandom(nNumberEvents, 1); 
vRatesCDF = [cumsum(vForecast(:, Forecast_Rate)./fTotalForecastRate)];

for nCnt = 1:nNumberEvents     (step 1)
  nIndex = min(find((vRandom(nCnt) < vRatesCDF)));       (step 2)
  vForecast(nIndex, Forecast_Observations) = vForecast(nIndex, Forecast_Observations) + 1;  (step 3)
  vForecast(nIndex, Forecast_PrecomputedZeroLikelihood) = calc_logpoisspdf(vForecast(nIndex, Forecast_Observations), vForecast(nIndex, Forecast_Rate)); (   step 4)
end;
fLikelihood = sum(vForecast(:, Forecast_PrecomputedZeroLikelihood)); (step 5)
--------------Code Ends-----------------------
This does not make sense all to me. If I am wrong, please guide me. 

However, I have found another implementation of Likelihood:
    Rhoades et al. 2011. (Equation 4): 
             A revised equation of LogLikelihood formula is provided. Which I think can be a bit efficient, in a way that it skips the use of a Log and a Factorial. Rest of all steps to be remained same. 

Please enlighten me, if I am not getting it right.
Regards.